server/src/main/java/org/elasticsearch/monitor/fs/FsHealthService.java

```diff
@@ -0,0 +1,179 @@
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.LongSupplier;
+
+/**
+ * Runs periodically and attempts to create a temp file to see if the filesystem is writable. If not then it marks the
+ * path as unhealthy.
+ */
+public class FsHealthService extends AbstractLifecycleComponent {
+
+    private static final Logger logger = LogManager.getLogger(FsHealthService.class);
+
+    private final ThreadPool threadPool;
+    private volatile boolean enabled;
+    private volatile TimeValue refreshInterval;
+    private volatile TimeValue healthcheckTimeoutInterval;
+    private final NodeEnvironment nodeEnv;
+    private final LongSupplier currentTimeMillisSupplier;
+    private AtomicLong lastSuccessfulRunTimeMillis = new AtomicLong(Long.MIN_VALUE);
+    private Map<Path, Status> pathHealthStats;
+    private volatile Scheduler.Cancellable scheduledFuture;
+
+    enum Status { HEALTHY, UNHEALTHY, UNKNOWN }
+
+    public static final Setting<Boolean> ENABLED_SETTING =
+        Setting.boolSetting("monitor.fs.health.enabled", true, Setting.Property.NodeScope, Setting.Property.Dynamic);
+    public static final Setting<TimeValue> REFRESH_INTERVAL_SETTING =
+        Setting.timeSetting("monitor.fs.health.refresh_interval", TimeValue.timeValueSeconds(5), TimeValue.timeValueSeconds(1),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+    public static final Setting<TimeValue> HEALTHCHECK_TIMEOUT_SETTING =
+        Setting.timeSetting("monitor.fs.health.unhealthy_timeout", TimeValue.timeValueMinutes(5), TimeValue.timeValueMinutes(2),
+            Setting.Property.NodeScope, Setting.Property.Dynamic);
+
+
+    FsHealthService(Settings settings, ClusterSettings clusterSettings, ThreadPool threadPool, NodeEnvironment nodeEnv,
+                    LongSupplier currentTimeMillisSupplier) {
+        this.threadPool = threadPool;
+        this.enabled = ENABLED_SETTING.get(settings);
+        this.refreshInterval = REFRESH_INTERVAL_SETTING.get(settings);
+        this.healthcheckTimeoutInterval = HEALTHCHECK_TIMEOUT_SETTING.get(settings);
+        this.currentTimeMillisSupplier = currentTimeMillisSupplier;
+        this.nodeEnv = nodeEnv;
+        this.pathHealthStats = new HashMap<>();
+        clusterSettings.addSettingsUpdateConsumer(REFRESH_INTERVAL_SETTING, this::setRefreshInterval);
+        clusterSettings.addSettingsUpdateConsumer(HEALTHCHECK_TIMEOUT_SETTING, this::setHealthcheckTimeoutInterval);
+        clusterSettings.addSettingsUpdateConsumer(ENABLED_SETTING, this::setEnabled);
+    }
+
+    @Override
+    protected void doStart() {
+        //TODO check if this needs to be a part of a dedicated threadpool
+        scheduledFuture = threadPool.scheduleWithFixedDelay(new FsHealthMonitor(), refreshInterval, ThreadPool.Names.SAME);
```

Comment:
I think this should not be on the `SAME` threadpool since it's doing IO that's potentially slow. `GENERIC` would be ok, but then I think we need protection to make sure there's only one check running at once.

Reply:
Done. Also modified the scheduled checks to be one per data path to honour the 1s `HEALTHY` SLA